src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -3312,6 +3319,12 @@ private enum PostSkipAction
             SyntaxKind expected,
             out GreenNode trailingTrivia)
         {
+            if (abortFunction(this))
+            {
+                trailingTrivia = null;
+                return PostSkipAction.Abort;
+            }
```

Comment:
this is a change to a pretty core function. can you explain what you're trying to do here and why it was necessary?

Reply:
Without this change the method will always return Continue if current token is expected (i.e. isNotExpectedFunction evaluates to false) and the same time current token marks a place when current parsing rule should be aborted (i.e. abortFunction returns true). This case appear in type argument list, when we expect comma or type, which is IdentifierToken, and should stop before member name, which is also IdentifierToken. So this change give higher priority in this situation to PostSkipAction.Abort. Otherwise caller will forever try to continue with current parsing rule, when actually reached a terminator. In all other cases this change won't influence the result of this method. 

Another option to solve this problem without modification of SkipBadTokensWithExpectedKind is to put IsTerminator() into condition in else if inside while loop in ParseTypeArgumentList:
```cs
private void ParseTypeArgumentList(out SyntaxToken open, SeparatedSyntaxListBuilder<TypeSyntax> types, out SyntaxToken close)
{
	Debug.Assert(this.CurrentToken.Kind == SyntaxKind.LessThanToken);
	open = this.EatToken(SyntaxKind.LessThanToken);
	open = CheckFeatureAvailability(open, MessageID.IDS_FeatureGenerics);

	if (this.IsOpenName())
	{
		// NOTE: trivia will be attached to comma, not omitted type argument
		var omittedTypeArgumentInstance = _syntaxFactory.OmittedTypeArgument(SyntaxFactory.Token(SyntaxKind.OmittedTypeArgumentToken));
		types.Add(omittedTypeArgumentInstance);
		while (this.CurrentToken.Kind == SyntaxKind.CommaToken)
		{
			types.AddSeparator(this.EatToken(SyntaxKind.CommaToken));
			types.Add(omittedTypeArgumentInstance);
		}

		close = this.EatToken(SyntaxKind.GreaterThanToken);

		return;
	}

	// first type
	types.Add(this.ParseTypeArgument());

	// remaining types & commas
	while (true)
	{
		if (this.CurrentToken.Kind == SyntaxKind.GreaterThanToken)
		{
			break;
		}
		else if (this.CurrentToken.Kind == SyntaxKind.CommaToken || this.IsPossibleTypeInTypeArgumentList())
		{
			types.AddSeparator(this.EatToken(SyntaxKind.CommaToken));
			types.Add(this.ParseTypeArgument());
		}
		else if (IsTerminator() || this.SkipBadTypeArgumentListTokens(types, SyntaxKind.CommaToken) == PostSkipAction.Abort)
		{
			break;
		}
	}

	close = this.EatToken(SyntaxKind.GreaterThanToken);
}

Reply:
> Another option to solve this problem without modification of SkipBadTokensWithExpectedKind is to put IsTerminator() into condition in else if inside while loop in ParseTypeArgumentList:

I would prefer that. 

it sounds like you're saying (paraphrased): "when we see an identifier we don't *always* want to consume it for the type parameter list, as that identifier might actually be the name of a member".    If that's hte case, i think it would be better to just edit the code that checks for type parameters, and make it so that it doens't just unilaterally assume an identifier is a type parameter (i.e. it also checks the surrounding tokens).

Reply:
Removed changes from SkipBadTokensWithExpectedKind. As you suggested, changed logic in SkipBadTypeArgumentListTokens, which now considers surrounding tokens, when checking for possible type.